tests/rptest/tests/datalake/datalake_e2e_test.py

```diff
@@ -952,3 +956,100 @@ def test_rest_catalog_metrics(self, cloud_storage_type):
+            },
+            *args,
+            **kwargs)
+        self.test_ctx = test_ctx
+
+    def setUp(self):
+        pass
+
+    def is_topic_fully_caught_up(self, topic_name: str):
+        admin = Admin(self.redpanda)
+        partitions = admin.get_partitions(topic=topic_name)
+
+        for p in partitions:
+            p_id = p['partition_id']
+            status = admin.get_partition_state("kafka",
+                                               topic=topic_name,
+                                               partition=p['partition_id'])
+            for replica in status['replicas']:
+                c_index = replica['raft_state']['commit_index']
+                for stm in replica['raft_state']['stms']:
+                    self.logger.debug(
+                        f"{topic_name}/{p_id} state machine: {stm['name']} on: {replica['raft_state']['node_id']} last_applied_offset: {stm['last_applied_offset']}"
+                    )
+                    if stm['last_applied_offset'] < c_index:
+                        return False
+
+        return True
+
+    @cluster(num_nodes=6)
+    # this test doesn't have to run with different catalog types
+    @matrix(cloud_storage_type=supported_storage_types(),
+            catalog_type=[CatalogType.REST_JDBC])
+    @skip_debug_mode
+    def test_enabling_iceberg_in_existing_cluster(self, cloud_storage_type,
+                                                  catalog_type):
+        count = 100
+        with DatalakeServices(self.test_ctx,
+                              redpanda=self.redpanda,
+                              include_query_engines=[QueryEngineType.SPARK],
+                              catalog_type=catalog_type) as dl:
+
+            topic = TopicSpec(name="delayed-iceberg-topic",
+                              partition_count=3,
+                              replication_factor=3,
+                              segment_bytes=1024 * 1024,
+                              redpanda_remote_read=False,
+                              redpanda_remote_write=False)
+
+            DefaultClient(dl.redpanda).create_topic(topic)
+
+            # produce ~60 MiB to the topic
+            dl.produce_to_topic(topic.name, 1024, 120 * 1024)
+            # waif for a while for the local snapshot to be taken
+            time.sleep(5)
+            dl.redpanda.restart_nodes(dl.redpanda.nodes)
+
+            def wait_for_topic(topic_name: str):
+                wait_until(lambda: self.is_topic_fully_caught_up(topic_name),
+                           timeout_sec=30,
+                           backoff_sec=1,
+                           err_msg=f"Error waiting for topic {topic_name} \
+                        state machines to catch up")
+
+            wait_for_topic(topic.name)
+            bytes_read_before = self.redpanda.estimate_bytes_read()
+
+            # enable iceberg, this will restart the cluster
+            dl.redpanda.set_cluster_config({"iceberg_enabled": True},
```

Comment:
Now that we have this new mechanism to avoid extra reads for Iceberg, should we remove the condition in the STM factory that disables the STM iceberg_enabled is false? The new approach seems more robust and doesn't suffer from the lack of determinism that the cluster property has

Reply:
let's do it in the follow up PR, ok ?